tiledb/cloud/notebook.py

```diff
@@ -35,3 +54,288 @@ def rename_notebook(
         )
     except GenApiException as exc:
         raise tiledb_cloud_error.check_exc(exc) from None
+
+
+def download_ipnyb_file_name_from_cloud(
+    tiledb_uri: str,
+    storage_credential_name: str,
+    ipynb_file_name: str,
+    *,
+    character_encoding: Optional[str] = 'utf-8',
+) -> None:
+    """
+    Downloads a notebook file from TileDB Cloud to local disk.
+    :param str tiledb_uri: such as "TileDB-Inc/quickstart_dense".
+    :param str storage_credential_name: such as "janedoe-creds", typically from the user's account settings.
+    :param str ipnyb_file_name: path to save to, such as "./mycopy.ipynb". Must be local; no S3 URI support at present.
+    :param str character_encoding: Optional character encoding of the file; defaults to utf-8 which is ASCII-compatible.
+    """
+    ipynb_file_contents = download_ipnyb_file_contents_from_cloud(
+        tiledb_uri,
+        storage_credential_name,
+        character_encoding=character_encoding,
+    )
+    # This can throw FileNotFoundError, PermissionError, etc.
+    # No need for try-except-raise only to re-throw the same exception.
+    with open (ipynb_file_name, "w") as handle:
+        handle.write(ipynb_file_contents)
+
+def download_ipnyb_file_contents_from_cloud(
+    tiledb_uri: str,
+    storage_credential_name: str,
+    *,
+    character_encoding: Optional[str] = 'utf-8',
+) -> str:
+    """
+    Downloads a notebook file from TileDB Cloud to contents as a string, nominally in JSON format.
+    :param str tiledb_uri: such as "TileDB-Inc/quickstart_dense".
+    :param str storage_credential_name: such as "janedoe-creds", typically from the user's account settings.
+    :param str character_encoding: Optional character encoding of the file; defaults to utf-8 which is ASCII-compatible.
+    :return: contents of the notebook file as a string, nominally in JSON format.
+    """
+    ctx = tiledb.cloud.Ctx({
+        "rest.creation_access_credentials_name": storage_credential_name,
+    })
+    with tiledb.open(tiledb_uri, 'r', ctx=ctx) as A:
+        size = A.meta["file_size"]
+        data = A.query(attrs=["contents"])[slice(0, size)]
+        json = data["contents"].tostring().decode("utf-8", "backslashreplace")
```

Comment:
This should probably be `tobytes()`, since `tostring()` is a deprecated alias.

I also don’t think we want a special error handler here; if the data we get back is invalid UTF-8, that is a problem that we probably don’t want to paper over.